### PR TITLE
Suppress console.log & others inside the TS Server runtime so that language service plugins don't accidentally use them and kill the server

### LINKED ISSUE #31209: TSServer: RangeError: Index out of range, exits with code: 134
Typescript 3.3.3333.  Haven't tried on @latest or @next yet because of that being a pending migration task - but since there is no matching report of this bug, I have low expectation that it's been fixed.

I searched for issues with:

* "RangeError: Index out of range"
* "TSServer exited with code: 134" but no hits.

My team is getting the following error:

```
[Info] Using tsserver from: ~\Repos\costar-suite\node_modules\typescript\lib
[Info] Forking TSServer
[Info] Started TSServer
[Error] ReaderError
RangeError: Index out of range
	at Buffer.copy (buffer.js:635:12)
	at l.tryReadContent (~\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:1:183067)
	at t.Reader.onLengthData (~\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:1:183633)
	at Socket.t.Reader.constructor.e.on.e (~\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\typescript-language-features\dist\extension.js:1:183400)
	at Socket.emit (events.js:182:13)
	at addChunk (_stream_readable.js:279:12)
	at readableAddChunk (_stream_readable.js:264:11)
	at Socket.Readable.push (_stream_readable.js:219:10)
	at Pipe.onread (net.js:636:20)
[Error] TSServer exited with code: 134
```

We were able to find this hint https://github.com/Microsoft/vscode/issues/71076 which suggested to me that tslint errors in custom tslint rules themselves can cause this issue.  Indeed, when I followed the only tslinter advice in a custom tslint rule code to replace a for-loop with a for-in loop, stability **greatly** improved.  It went from crashing repeatedly and continuously to crashing less frequently - after about an hour of use.  But the error output is the same.

If there are additional diagnostics that I can run to help pinpoint this issue, I will try it.  If this is known as a dupe and/or already fixed, sorry for the noise.

### PR DESCRIPTION
**Pull request by orta**
Fixes #31209 (where most of the discussion is)

This should probably hold till 3.7

### CODE REVIEW

**Comment by weswigham** [comment]
@RyanCavanaugh would it be possible to just use a private fd (spawning a process with extra pipes established is [pretty easy](https://nodejs.org/api/child_process.html#child_process_options_stdio) in node) rather than stdout in most tsserver environments? We already use a socket for events... it should be a change to `server.ts` to not always make the readline instance from `process.stdin` and `process.stdout`.

File: src/tsserver/server.ts [hunk 1]
```diff
@@ -970,4 +970,12 @@ namespace ts.server {
     if (ts.sys.tryEnableSourceMapsForHost && /^development$/i.test(ts.sys.getEnvironmentVariable("NODE_ENV"))) {
         ts.sys.tryEnableSourceMapsForHost();
     }
+
+    // Overwrites the current console messages to instead write to
+    // the log. This is so that language service plugins which use
+    // console.log don't break the message passing between tsserver
+    // and the client
+    console.log = (msg) => logger.msg(msg, Msg.Info);
```
Comment (nmain) [inline src/tsserver/server.ts]: console.log/warn/error are normally variadic in node, so would it make sense to do something with all `...args` in this replacement function?  We'd be less likely to lose important information if a plugin passed multiple things to one log call.

Comment (orta) [reply src/tsserver/server.ts]: I don't think the logger accepts variadic args - we can pass the array joined as a string if args.length is greater than 1 though.

**Comment by orta** [comment]
We chatted about this and the idea of using a private file descriptor for tsserver comms would be a good move, but that requires all tsserver editors making changes to handle that instead of stdin/stdout - so this should be a good enough for now.

**Pull request merged**